fem/field_interpolant.hpp

```diff
@@ -0,0 +1,191 @@
+// Copyright (c) 2010-2020, Lawrence Livermore National Security, LLC. Produced
+// at the Lawrence Livermore National Laboratory. All Rights reserved. See files
+// LICENSE and NOTICE for details. LLNL-CODE-806117.
+//
+// This file is part of the MFEM library. For more information and source code
+// availability visit https://mfem.org.
+//
+// MFEM is free software; you can redistribute it and/or modify it under the
+// terms of the BSD-3 license. We welcome feedback and contributions, see file
+// CONTRIBUTING.md for details.
+
+// Implementation of Field Interpolants
+
+#ifndef MFEM_FIELD_INTERPOLANT
+#define MFEM_FIELD_INTERPOLANT
+
+#include "../config/config.hpp"
+#include "../linalg/linalg.hpp"
+#include "intrules.hpp"
+#include "eltrans.hpp"
+#include "coefficient.hpp"
+#include "bilininteg.hpp"
+#include "bilinearform.hpp"
+#include "lininteg.hpp"
+#include "gridfunc.hpp"
+#ifdef MFEM_USE_MPI
+#include "pgridfunc.hpp"
+#include "pbilinearform.hpp"
+#endif
+
+namespace mfem
+{
+
+class FieldInterpolant
```

Comment:
- I think that the methods in this class should be methods of `(Par)GridFunction`. They are specific versions of `ProjectCoefficient`. Maybe we should discuss this.

- The constructor is very confusing, as it creates mass matrices (for some of the functions) on one space, while the methods use another space.

- Should be noted that these don't work for `non-VectorFE` spaces.

- Even if this class stays, the name is too general. The name should somehow state  `Quadrature`.

- All comments about `public` stuff should be converted to doxygen.

- I'm confused by the 'Disc' version. First, it's probably not needed, because the other versions should work for both CG and DG (it would just invert a global matrix instead of local ones). Second, the existing `GridFunction::ProjectDiscCoefficient` projects a `Coefficient` with jumps into a **CG** function. This projects a Coefficient into a **DG** function. So the name doesn't correspond to the old usage.

Reply:
I'll comment on these in order of the bullet points.
1. The ```ProjectQuadratureDiscCoefficient``` could be moved over there especially now that they are essentially wrappers around ```GridFunction::ProjectDiscCoefficient```. The wrapper just adds a few more checks to make sure that ```IntegrationRules``` are the same between the two. I would like to move the ```ProjectQuadratureCoefficient``` over there as well. However, this method requires the use of a ```BilinearForm``` which depend on ```GridFunction```, so I wanted to avoid the possibility of dealing with some sort of circular references in those sections.
2. This should hopefully be a bit nicer/less confusing, since none of the functions now have you pass in a FES. The space used with the mass matrix and those found on the gridfunctions passed to ```ProjectQuadratureCoefficient``` should be the same. I currently don't know of a way to check that so if you're aware of a way I would like some help there.
3. I originally wanted something like QuadratureInterpolant but that was already taken so I went with the reverse. I've now changed it to ```Quad2FieldInterpolant```. I did consider ```Quadrature2FieldInterpolant``` but that seemed too long.
4. I believe this should now be fixed. If you catch anything that appears off or was missed let me know.
5. Some of this was remnants from trying to set something that could get us onto the GPU later on in the future. After some of the discussions as seen above, I realized it could go the way I had originally or just use the ```ProjectDiscCoefficient``` method with a few verifications added to it. I've now swapped over to  ```ProjectDiscCoefficient``` method. Although, I do believe we might be able to eventually use something like I originally had within the ```ProjectDiscCoefficient``` along with @artv3 PR on Batch LU to get us onto the GPU for at least L2 and H1 spaces for that method.